File: common/src/main/java/com/ssblur/alchimiae/recipe/MashRecipe.java
package com.ssblur.alchimiae.recipe;

import com.ssblur.alchimiae.alchemy.AlchemyHelper;
import com.ssblur.alchimiae.item.AlchimiaeItems;
import net.minecraft.core.HolderLookup;
import net.minecraft.core.NonNullList;
import net.minecraft.core.component.DataComponents;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.CraftingBookCategory;
import net.minecraft.world.item.crafting.CraftingInput;
import net.minecraft.world.item.crafting.CustomRecipe;
import net.minecraft.world.item.crafting.RecipeSerializer;
import net.minecraft.world.level.Level;

public class MashRecipe extends CustomRecipe {
  ServerLevel server;
  public MashRecipe(CraftingBookCategory craftingBookCategory) {
    super(craftingBookCategory);
  }

  @Override
  public boolean matches(CraftingInput recipeInput, Level level) {
    if(recipeInput.items().stream().noneMatch(item -> item.is(AlchimiaeItems.GRINDER)))
      return false;
    if(recipeInput.ingredientCount() <= 2)
      return false;

    if(level instanceof ServerLevel serverLevel) {
      this.server = serverLevel;
      var items = recipeInput.items().stream().filter(item -> !item.is(AlchimiaeItems.GRINDER)).toList();

      return AlchemyHelper.getEffects(items, server) != null;
    } else {
      return false;
    }
  }

  @Override
  public ItemStack assemble(CraftingInput recipeInput, HolderLookup.Provider provider) {
    var output = new ItemStack(AlchimiaeItems.MASH);
    if(server == null || !server.getServer().isRunning()) return output;
    var items = recipeInput.items().stream().filter(item -> !item.is(AlchimiaeItems.GRINDER)).toList();
    output.set(
      DataComponents.POTION_CONTENTS,
      AlchemyHelper.getPotionContents(items, server)
    );
    return output;
  }

  @Override
  public boolean canCraftInDimensions(int i, int j) {
    return i + j >= 1;
  }

  @Override
  public RecipeSerializer<?> getSerializer() {
    return AlchimiaeRecipes.MASH.get();
  }

  @Override
  public NonNullList<ItemStack> getRemainingItems(CraftingInput recipeInput) {
    NonNullList<ItemStack> list = NonNullList.withSize(recipeInput.size(), ItemStack.EMPTY);
    for(int i = 0; i < recipeInput.size(); i++) {
      var item = recipeInput.getItem(i);
      if(item.is(AlchimiaeItems.GRINDER) && item.isDamageableItem()) {
        item.setDamageValue(item.getDamageValue() + 1);
        if(item.getDamageValue() >= item.getMaxDamage())
          item = ItemStack.EMPTY;
        list.set(i, item.copy());
      }
    }
    return list;
  }
}

File: common/src/main/kotlin/com/ssblur/alchimiae/recipe/MashRecipe.kt
package com.ssblur.alchimiae.recipe

import com.ssblur.alchimiae.alchemy.AlchemyHelper
import com.ssblur.alchimiae.data.AlchimiaeDataComponents
import com.ssblur.alchimiae.item.AlchimiaeItems
import com.ssblur.alchimiae.item.GrinderItem
import net.minecraft.core.HolderLookup
import net.minecraft.core.NonNullList
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.crafting.CraftingBookCategory
import net.minecraft.world.item.crafting.CraftingInput
import net.minecraft.world.item.crafting.CustomRecipe
import net.minecraft.world.item.crafting.RecipeSerializer
import net.minecraft.world.item.enchantment.EnchantmentHelper
import net.minecraft.world.item.enchantment.Enchantments
import net.minecraft.world.level.Level
import kotlin.random.Random

class MashRecipe(craftingBookCategory: CraftingBookCategory?) : CustomRecipe(craftingBookCategory) {
  var server: ServerLevel? = null
  override fun matches(recipeInput: CraftingInput, level: Level): Boolean {
    if (recipeInput.items().stream().noneMatch { item: ItemStack -> item.`is`(AlchimiaeItems.GRINDER) }) return false
    if (recipeInput.ingredientCount() <= 2) return false

    if (level is ServerLevel) {
      this.server = level
      val items = recipeInput.items().stream().filter { item: ItemStack -> !item.`is`(AlchimiaeItems.GRINDER) }.toList()

      return AlchemyHelper.getEffects(items, server!!, 1.0f) != null
    } else {
      return false
    }
  }

  override fun assemble(recipeInput: CraftingInput, provider: HolderLookup.Provider): ItemStack {
    val output = ItemStack(AlchimiaeItems.MASH)
    if (server == null || !server!!.server.isRunning) return output
    var efficiency = 0.5f
    val grinder = recipeInput.items().stream().filter { item: ItemStack -> item.`is`(AlchimiaeItems.GRINDER) }.findAny()
    if (grinder.isPresent && grinder.get().item is GrinderItem)
      efficiency = (grinder.get().item as GrinderItem).efficiency
    val items = recipeInput.items().stream().filter { item: ItemStack -> !item.`is`(AlchimiaeItems.GRINDER) }.toList()
    output.set(
      AlchimiaeDataComponents.CUSTOM_POTION,
      AlchemyHelper.getCustomPotionEffects(items, server!!, efficiency)
    )
    return output
  }

  override fun canCraftInDimensions(i: Int, j: Int): Boolean {
    return i + j >= 1
  }

  override fun getSerializer(): RecipeSerializer<*> {
    return AlchimiaeRecipes.MASH.get()
  }

  override fun getRemainingItems(recipeInput: CraftingInput): NonNullList<ItemStack> {
    val list = NonNullList.withSize(recipeInput.size(), ItemStack.EMPTY)
    for (i in 0..<recipeInput.size()) {
      var item = recipeInput.getItem(i).copy()
      var unbreaking = 0
      EnchantmentHelper.getEnchantmentsForCrafting(item).entrySet().forEach {
        if(it.key == Enchantments.UNBREAKING)
          unbreaking += it.intValue
      }

      if (item.`is`(AlchimiaeItems.GRINDER) && item.isDamageableItem) {
        if(Random.nextInt(10) > unbreaking) item.damageValue = item.damageValue + 1
        if (item.damageValue >= item.maxDamage) item = ItemStack.EMPTY
        list[i] = item
      }
    }
    return list
  }
}
